src/commonMain/kotlin/app/cash/turbine/channel.kt

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.turbine
+
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.channels.ChannelResult
+import kotlinx.coroutines.channels.ClosedReceiveChannelException
+import kotlinx.coroutines.channels.ReceiveChannel
+
+/**
+ * Returns the most recent item that has already been received.
+ * If channel was closed with no item being received
+ * previously, this function will throw an [AssertionError]. If channel
+ * was closed with an exception, this function will throw the underlying exception.
+ *
+ * @throws AssertionError if no item was emitted.
+ */
+public fun <T> ReceiveChannel<T>.expectMostRecentItem(): T {
+  var result: ChannelResult<T>? = null
+  var prevResult: ChannelResult<T>?
+  while (true) {
+    prevResult = result
+    result = tryReceive()
+    result.exceptionOrNull()?.let { throw it }
+    if (!result.isSuccess) {
+      break
+    }
+  }
+
+  if (prevResult?.isSuccess == true) return prevResult!!.getOrThrow()
+
+  throw TurbineAssertionError("No item was found", cause = null)
+}
+
+/**
+ * Assert that there are no unconsumed events which have already been received.
+ *
+ * A channel in the closed state will always emit either [Event.Complete] or [Event.Error] when read, so
+ * [expectNoEvents] will only succeed on an empty [ReceiveChannel] that is not closed.
+ *
+ * @throws AssertionError if unconsumed events are found.
+ */
+public fun <T> ReceiveChannel<T>.expectNoEvents() {
+  val result = tryReceive()
+  if (result.isSuccess || result.isClosed) result.unexpectedResult("no events")
+}
+
+/**
+ * Assert that an event was received and return it.
+ * This function will suspend if no events have been received.
+ *
+ * This function will always return a terminal event on a closed [ReceiveChannel].
+ */
+public suspend fun <T> ReceiveChannel<T>.awaitEvent(): Event<T> =
+  try {
+    Event.Item(receive())
+  } catch (e: CancellationException) {
+    throw e
+  } catch (e: ClosedReceiveChannelException) {
+    Event.Complete
+  } catch (e: Exception) {
+    Event.Error(e)
+  }
+
+/**
+ * Assert that the next event received was non-null and return it.
+ * This function will not suspend, and will throw if invoked in a suspending context.
```

Comment:
I think we need to add "on the JVM or Android" to all these statements otherwise or else use terminology like "may" or "attempt" to indicate that it's not reliable on all supported platforms.